src/NUnitFramework/framework/Constraints/UniqueItemsConstraint.cs

```diff
@@ -174,5 +192,39 @@ private Type GetGenericTypeArgument(IEnumerable actual)
 
             return null;
         }
+
+        private sealed class CaseInsensitiveCharComparer : IEqualityComparer<char>
+        {
```

Comment:
I can't find a way to do a case-insensitive culture-aware char comparison. I'd been hoping to avoid converting each char in the collection to a 1-length string

Reply:
@jnm2 Do you have input on this. You usually have a good grab regarding case-insensitive comparison and best practises

Reply:
`StringComparer.CurrentCultureIgnoreCase`?

Reply:
So here's the problem: culture-sensitive comparisons are broken when you make them operate on individual chars. How is `"ß".Equals("ss", StringComparison.CurrentCulture)` being true going to be reflected when you aren't passing the whole string for the comparison?

Can we get out of comparing a char at a time?

Reply:
In situations where operating on individual paired chars makes sense (i.e. where `x.ToString().Equals(y.ToString(), ...)` would be a valid thing to do), .NET Core 2.1 adds an Equals extension method on ReadOnlySpan<char> instances with StringComparison. That lets you avoid the .ToString() allocation. But as mentioned above, operating on individual chars is probably not a valid way to be doing any kind of linguistic comparison.

Reply:
One last twist is that `char` in .NET doesn't represent what you'd think of as a character. It's UTF-16, it either represents one "character" or half of a "character" (surrogate pairs).

Reply:
Ah ok, thanks for explaining! I must admit, I don't have much experience with internationalization at application level. I'll push a few updates over the next little bit to try and make this more robust

Reply:
I've done some digging and it seems like I've run into an existing bug in that the `StringsComparer` is [not culture-safe](https://github.com/nunit/nunit/blob/master/src/NUnitFramework/framework/Constraints/Comparers/StringsComparer.cs#L48-L53), as demonstrated by the below:

```csharp
[Test, SetCulture("de-DE")]
        public void CultureAware()
        {
            string x = "ss";
            string y = "ß";

            Assert.IsTrue(string.Equals(x, y, StringComparison.CurrentCulture));  // passes
            Assert.AreEqual(x, y);   // fails
        }
```

This same issue is causing failures for the medium/slow paths when the framework has to fall back to using the NUnitEqualityComparer to check for uniques:
https://github.com/nunit/nunit/pull/3523/files#diff-a31a83d72a32eb0e9538e5ca251a3876R100-R113

I can fix this easy enough by updating StringsComparer to use `CurrentCulture/CurrentCultureIgnoreCase`, but it seems we have an existing test which relies on using an OrdinalComparison. This test will fail if run using either `CurrentCulture/CurrentCultureIgnoreCase`:
https://github.com/nunit/nunit/blob/ad943eb/src/NUnitFramework/tests/Assertions/StringAssertTests.cs#L155-L162

What would the best approach here be?

Reply:
Eesh, this has turned into a can of worms, hasn't it! What do you think @jnm2 - 'fix' it to be culture sensitive, and document the change? At least that would be more consistent across the library.

That would mean to specifically compare for different encodings, you'd have to specifically put in an Ordinal string comparer. That seems sensible, but is of course a change in behaviour...

Reply:
Definitely a good exercise in corner cases. 😄 

While digging a bit deeper, I've also noticed that the "different encoding" behaviour itself is a bit inconsistent on the current `master` branch as the test only passes when `IgnoreCase` is false.

```csharp
[Test]
public void DifferentEncodingsOfSameStringAreNotEqual()
{
    string input = "Hello World";
    byte[] data = System.Text.Encoding.Unicode.GetBytes( input );
    string garbage = System.Text.Encoding.UTF8.GetString( data, 0, data.Length);

    Assert.That(input, Is.Not.EqualTo(garbage));             // Passes
    Assert.That(input, Is.Not.EqualTo(garbage).IgnoreCase);  // Fails
}
```

Reply:
It looks like this encoding test is actually quite a long-standing change (going back to [NUnit 2.5.1](https://github.com/nunit/nunitv2/commit/641f0e14631a62ab7fe138ed3618a6c99a888def), and referencing a bug tracker ID I'm not familiar with). I'm unsure when case insensitivity was added in relation to it, but I'm thinking it might be worth rolling back all culture-aware changes made in this PR (and going back to using `ToLower()`) and instead spinning off a separate ticket to discuss explicitly moving to `CurrentCulture/CurrentCultureIgnoreCase`.

Thoughts?

Reply:
Would that be a lot of work? I think that's possibly a good idea...the culture issue doesn't really feel like one to be tackled bit-by-bit. 😢

Reply:
Yes, let's try to keep the issue separate from this PR so long as we aren't regressing the master branch in any way.

Reply:
Done (at least as far as this branch goes). Will make an issue shortly

Reply:
I've created https://github.com/nunit/nunit/issues/3549

Reply:
I agree with @jnm2 let's keep all this separate. Thanks for creating the issue @stevenaw. I'm going to mark this conversations resolved.